### PR TITLE
[puller] download post-optin

### PR DESCRIPTION
**Pull request by gimre-xymcity**
I'm missing client.redeem test, but please take a look

### CODE REVIEW

File: optin/puller/puller/client/SymbolClient.py [hunk 1]
```diff
@@ -66,6 +66,16 @@ async def transaction_statuses(self, transaction_hashes):
 		request = {'hashes': [str(transaction_hash) for transaction_hash in transaction_hashes]}
 		return await self.post('transactionStatus', request)
 
+	async def redeem_transactions(self, optin_signer_public_key, start_id=None):
```
Comment (Jaguar0625) [inline optin/puller/puller/client/SymbolClient.py:69]: find_redeemed_transactions ?

File: optin/puller/puller/client/SymbolClient.py [hunk 2]
```diff
@@ -66,6 +66,16 @@ async def transaction_statuses(self, transaction_hashes):
...
+		"""Finds redeem transactions"""
```
Comment (Jaguar0625) [inline optin/puller/puller/client/SymbolClient.py:70]: missing period

File: optin/puller/workflows/generate_postoptin.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+def parse_args():
+	parser = argparse.ArgumentParser(description='generate postoptin json')
+	parser.add_argument('--symbol-node', help='Symbol node url', default='http://wolf.importance.jp:3000')
+	parser.add_argument('--optin-signer-public-keys', help='optin signer public keys (comma separated)', default=OPTIN_SIGNER_PUBLIC_KEYS)
+	parser.add_argument('--output', help='output path', default='post-optin.json')
+	return parser.parse_args()
+
+
+def to_sym_address(hex_string):
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:24]: sym => symbol

File: optin/puller/puller/client/SymbolClient.py [hunk 4]
```diff
@@ -66,6 +66,16 @@ async def transaction_statuses(self, transaction_hashes):
...
+
+		url_path = f'transactions/confirmed?signerPublicKey={optin_signer_public_key}&embedded=true&fromHeight=2&pageSize=100'
+		if start_id:
+			url_path += f'&offset={start_id}'
+
+		transactions = await self.get(url_path, 'data')
+		return transactions
+
 	async def find_optin_transactions(self, optin_signer_public_key, address):
```
Comment (Jaguar0625) [inline optin/puller/puller/client/SymbolClient.py:79]: (missing tests)

File: optin/puller/workflows/generate_postoptin.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+				break
+
+			for transaction in transactions:
+				self.process_transaction(transaction)
+
+			start_id = transactions[-1]['id']
+			print('.', end='', flush=True)
+
+	def process_transaction(self, transaction_meta):
+		transaction = transaction_meta['transaction']
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:47]: pass in transaction_meta['transaction'] ?

File: optin/puller/workflows/generate_postoptin.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		if TransactionType.TRANSFER.value != transaction['type']:
+			return
+
+		# note: ignoring txes without message
+
+		if 'message' in transaction:
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:53]: move to previous if

File: optin/puller/workflows/generate_postoptin.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+			message = unhexlify(transaction['message']).decode('utf8')
+			assert message[0] == '\0'
+
+			# "new" format
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:57]: hm?

File: optin/puller/workflows/generate_postoptin.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+			xym_amount = int(transaction['mosaics'][0]['amount'])
+			if message[1] == '{':
+				data = json.loads(message[1:])
+
+				nis_address = data.get('nisAddress', None)
+				if ' n1sAddress ' in data:
+					nis_address = data[' n1sAddress ']
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:64]: this was a valid one? 😅

File: optin/puller/workflows/generate_postoptin.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		await processor.process(public_key)
+
+		all_redemptions.extend(processor.redemptions)
+
+	with open(Path(args.output), 'wt', encoding='utf8') as out_file:
+		out_file.write(json.dumps(all_redemptions, indent=2))
+
+
+if '__main__' == __name__:
+	asyncio.run(main())
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:102]: how/where are you handling duplicate?

File: optin/puller/workflows/generate_postoptin.py (hunk 7 above)
Comment (gimre-xymcity) [reply optin/puller/workflows/generate_postoptin.py:57]: 'new' format is the one in json `{'nisaddress':'...'}`

File: optin/puller/workflows/generate_postoptin.py (hunk 8 above)
Comment (gimre-xymcity) [reply optin/puller/workflows/generate_postoptin.py:64]: 😑

File: optin/puller/puller/client/SymbolClient.py [hunk 10]
```diff
@@ -66,6 +66,16 @@ async def transaction_statuses(self, transaction_hashes):
 		request = {'hashes': [str(transaction_hash) for transaction_hash in transaction_hashes]}
 		return await self.post('transactionStatus', request)
 
+	async def find_redeem_transactions(self, optin_signer_public_key, start_id=None):
+		"""Finds redeem transactions."""
```
Comment (Jaguar0625) [inline optin/puller/puller/client/SymbolClient.py:70]: isn't "redeemed" more correct ?

File: optin/puller/workflows/generate_postoptin.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		self.redemptions = []
+
+	async def process(self, public_key):
+		start_id = None
+		while True:
+			transactions = await self.client.find_redeem_transactions(public_key, start_id)
+			if not transactions:
+				break
+
+			for transaction in transactions:
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:41]: probably would use more verbose transaction_and_metadata_pair(s)

File: optin/puller/workflows/generate_postoptin.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+				self.process_transaction(transaction['transaction'])
+
+			start_id = transactions[-1]['id']
+			print('.', end='', flush=True)
+		print()
+
+	def process_transaction(self, transaction):
+		# note: ignoring txes without message
+		if TransactionType.TRANSFER.value != transaction['type'] or not 'message' in transaction:
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:50]: lint

File: optin/puller/tests/client/test_SymbolClient.py [hunk 13]
```diff
@@ -268,9 +268,36 @@ async def test_can_query_transaction_statuses(server):  # pylint: disable=redefi
...
+	# Act:
+	transactions = await client.find_redeem_transactions(PublicKey(PUBLIC_KEYS[0]))
+
+	# Assert:
+	assert [
+		f'{server.make_url("")}/transactions/confirmed?signerPublicKey={PublicKey(PUBLIC_KEYS[0])}&embedded=true&fromHeight=2&pageSize=100'
+	] == server.mock.urls
+	assert 3 == len(transactions)
+
+	prefix = '007b20226e697341646472657373223a20224e'
```
Comment (Jaguar0625) [inline optin/puller/tests/client/test_SymbolClient.py:293]: uppercase 😅

Comment (gimre-xymcity) [reply optin/puller/tests/client/test_SymbolClient.py:293]: so I'd have to throw additional upper in mock server (feels weird to do it just for test purposes)

File: optin/puller/puller/client/SymbolClient.py (hunk 10 above)
Comment (Jaguar0625) [reply optin/puller/puller/client/SymbolClient.py]: they have already been redeemed, haven't they? or what are you trying to say?

File: optin/puller/workflows/generate_postoptin.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+from pathlib import Path
+
+from symbolchain.CryptoTypes import PublicKey
+from symbolchain.sc import TransactionType
+from symbolchain.symbol.Network import Address
+
+from puller.client.SymbolClient import SymbolClient
+
+OPTIN_SIGNER_PUBLIC_KEYS = '7AEC08AA66CB50B0C3D180DE7508D5D82ECEDCDC9E73F61FA7069868BEABA856'
+DUPLICATE_DESTINATION = ''
```
Comment (Jaguar0625) [inline optin/puller/workflows/generate_postoptin.py:14]: seems unused

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**